contents/newsletter/feature-flag-mistakes.md

```diff
@@ -0,0 +1,172 @@
+
+1. **Safety.** Flags enable progressive rollouts and kill switch rollbacks enabling you to deploy and release new code safely.
+
+2. **Speed.** By [separating deployment from release](/product-engineers/decouple-deployment-from-release), they enable us to do [trunk-based development](/product-engineers/trunk-based-development) meaning we can [ship more changes faster](https://newsletter.posthog.com/p/how-to-design-your-company-for-speed). 
+
+On top of this, they let us run [A/B tests](/experiments), [beta programs](/tutorials/public-beta-program), and [dogfood](/product-engineers/dogfooding) our changes. As a [remote and async team](https://newsletter.posthog.com/p/how-we-work-asynchronously), this is critical to our success. Our tactics for developing and releasing software would be a lot more limited without them.
+
+## 2. Mixing business and flag logic
+
+Structuring your usage of flags well is critical to using them. One of the biggest ways this can go wrong is by mixing flag logic with business logic.
+
+This means having feature flag checks in your application code. For example, if you wanted to roll out a new pricing algorithm with a feature flag, a bad way structure would look like this: 
+
+```js
+// Bad mixed logic
+function calculatePrice(product) {
+  let price = product.basePrice;
+  
+  if (posthog.isFeatureEnabled('new_pricing_algorithm')) {
+    price = price * 1.1;
+  }
+  
+  if (posthog.isFeatureEnabled('holiday_discount')) {
+    price = price * 0.8;
+  }
+  
+  if (posthog.isFeatureEnabled('member_pricing')) {
+    if (user.isMember) {
+      price = price * 0.9;
+    }
+  }
+
+  return price;
+}
+```
+
+This structure is difficult to test, difficult to understand the core logic, and easy to break the functionality if/when the flag changes. It is better to separate the pricing logic from the feature flag logic like this:
+
+```js
+// Good separated logic
+function calculatePrice(product, pricingStrategy) {
+  return pricingStrategy.calculatePrice(product);
+}
+
+// Flag logic separated into configuration/initialization
+const pricingStrategy = posthog.isFeatureEnabled('new_pricing_algorithm') 
+  ? new NewPricingStrategy()
+  : new StandardPricingStrategy();
+```
+
+An ideal structure would use the flag in as few places as possible and abstract as much as possible behind a single flag. This creates a clear separation and isolation of flag-related code which limits complexity and improves maintainability.
+
+Good goals for avoiding this mistake are only using the feature flag in a single spot and making it easy to remove. Accomplishing both reveals your code is structured well. 
+
+## 3. Being surprised when flags are turned off
+
+The core of a feature flag's value to developers is the ability to turn them off. They need to be ready for this at any point, which isn't always the case.
+
+![Shocked by feature flag](https://res.cloudinary.com/dmukukwp6/image/upload/turnoff_941966283f.png)
```

Comment:
I think this would be a good one to turn into a Maxhog meme – I'm sure Lottie could dress Max in the pikachu costume! We could use this as the main image.

Reply:
Requested: https://github.com/PostHog/posthog.com/issues/10321